lib/SILOptimizer/Transforms/DeadStoreElimination.cpp

```diff
@@ -888,8 +890,13 @@ bool DSEContext::processWriteForDSE(BlockState *S, unsigned bit) {
       continue;
     // If 2 locations may alias, we can still keep both stores.
     LSLocation &L = LocationVault[i];
-    if (!L.isMustAliasLSLocation(R, AA))
+    if (!L.isMustAliasLSLocation(R, AA)) {
+      // If this is a reference type, then the *live* store counts as an unkown
+      // read.
+      if (L.getBase()->getType().isAnyClassReferenceType())
+        S->stopTrackingLocation(S->BBWriteSetMid, i);
```

Comment:
This was always an issue. We just never ran into it because before this patch we struggled with reference types.